acsys/__init__.py

```diff
@@ -720,7 +729,21 @@ def handler(rpy, last):
 
             done = False
             while not done:
-                snd, sts, msg, done = await rpy_q.get()
+                next_reply = rpy_q.get()
+                reply_task = asyncio.create_task(next_reply)
+                pending.add(reply_task)
+
+                (resolved, pending) = await asyncio.wait(pending, return_when=asyncio.FIRST_COMPLETED)
+
+                # If `next_reply` is not in the resolved set, then the
+                # signalling future was set. In this case, the
+                # generator exits.
+
+                if not reply_task in resolved:
+                    return None
```

Comment:
We can't `return` inside a generator. Initial testing of `raise GeneratorExit` appears to work.